sources/api/netdog/src/net_config/mod.rs

```diff
@@ -0,0 +1,303 @@
+//! The net_config module contains the strucures needed to deserialize a `net.toml` file.  It also
+//! includes contains the `FromStr` implementations to create a `NetConfig` from string, like from
+//! the kernel command line.
+//!
+//! These structures are the user-facing options for configuring one or more network interfaces.
+mod dhcp;
+mod error;
+mod v1;
+
+use crate::wicked::WickedInterface;
+pub(crate) use dhcp::{Dhcp4ConfigV1, Dhcp4OptionsV1, Dhcp6ConfigV1, Dhcp6OptionsV1};
+pub(crate) use error::{Error, Result};
+use serde::Deserialize;
+use snafu::{ensure, ResultExt};
+use std::fs;
+use std::path::Path;
+use std::str::FromStr;
+pub(crate) use v1::NetConfigV1;
+
+static DEFAULT_INTERFACE_PREFIX: &str = "netdog.default-interface=";
+
+/// This trait must be implemented by each new version of network config
+pub(crate) trait Interfaces {
+    /// Returns the primary network interface.
+    fn primary_interface(&self) -> Option<String>;
+
+    /// Does the config contain any interfaces?
+    fn has_interfaces(&self) -> bool;
+
+    /// Converts the network config into a list of `WickedInterface` structs, suitable for writing
+    /// to file
+    fn into_wicked_interfaces(&self) -> Vec<WickedInterface>;
+}
+
+impl<I: Interfaces> Interfaces for Box<I> {
+    fn primary_interface(&self) -> Option<String> {
+        (**self).primary_interface()
+    }
+
+    fn has_interfaces(&self) -> bool {
+        (**self).has_interfaces()
+    }
+
+    fn into_wicked_interfaces(&self) -> Vec<WickedInterface> {
+        (**self).into_wicked_interfaces()
+    }
+}
+
+/// This private trait must also be implemented by each new version of network config.  It is used
+/// during the deserialization of the config to validate the configuration, ensuring there are no
+/// conflicting options set, etc.
+trait Validate {
+    /// Validate the network configuration
+    fn validate(&self) -> Result<()>;
+}
+
+impl<V: Validate> Validate for Box<V> {
+    fn validate(&self) -> Result<()> {
+        (**self).validate()
+    }
+}
+
+/// Read the network config from file, returning an object that implements the `Interfaces` trait
+pub(crate) fn from_path<P>(path: P) -> Result<Option<Box<dyn Interfaces>>>
+where
+    P: AsRef<Path>,
+{
+    let path = path.as_ref();
+    let net_config_str =
+        fs::read_to_string(path).context(error::NetConfigReadFailedSnafu { path })?;
+    let net_config = deserialize_config(&net_config_str)?;
+
+    if !net_config.has_interfaces() {
+        return Ok(None);
+    }
+
+    Ok(Some(net_config))
+}
+
+/// Deserialize the network config, using the version key to determine which config struct to
+/// deserialize into
+fn deserialize_config(config_str: &str) -> Result<Box<dyn Interfaces>> {
+    #[derive(Debug, Deserialize)]
+    struct ConfigToml {
+        version: u8,
```

Comment:
Is version as a field mandatory? Can we assume it's the latest version of net_config if it's not specified?

Reply:
Version is mandatory currently, to keep things explicit.

Some considerations:
* In order to know what "latest" is at any given time we'd need to maintain a constant or something that would we would need to remember to change.
* Without the version field, we'd need to attempt to deserialize into a struct with a version field first, and if that failed for some reason we'd attempt to deserialize into "latest"?

In this case I think I'd rather be explicit and avoid any guessing. :)